bundles/org.openhab.binding.kermi/src/main/java/org/openhab/binding/kermi/internal/api/KermiHttpUtil.java

```diff
@@ -0,0 +1,185 @@
+
+import com.google.gson.Gson;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+
+import io.micrometer.core.instrument.util.StringUtils;
+
+/**
+ * @author Marco Descher - Initial contribution
+ */
+public class KermiHttpUtil {
+
+    private static final String CONTENT_TYPE = "application/json; charset=utf-8";
+
+    private final Logger logger = LoggerFactory.getLogger(KermiHttpUtil.class);
+
+    private String hostname = "";
+    private String password = "";
+    private Properties httpHeaders;
+    private Gson gson;
+
+    public KermiHttpUtil() {
+        httpHeaders = new Properties();
+        gson = new Gson();
+    }
+
+    private String getBaseApiUrl() {
+        return "http://" + hostname + "/api/";
+    }
+
+    public void setHostname(String hostname) {
+        this.hostname = hostname;
+    }
+
+    public void setPassword(String password) {
+        this.password = password;
+    }
+
+    /**
+     * Issue a HTTP GET request and retry on failure
+     *
+     * @param url the url to execute
+     * @param timeout the socket timeout in milliseconds to wait for data
+     * @return the response body
+     * @throws KermiCommunicationException when the request execution failed or interrupted
+     */
+    @SuppressWarnings("null")
+    public synchronized String executeUrl(String httpMethod, String url, String content, String contentType)
+            throws KermiCommunicationException {
+        if (StringUtils.isBlank(hostname)) {
+            return "Not connected";
+        }
+
+        int attemptCount = 1;
+        try {
+            while (true) {
+                Throwable lastException = null;
+                String result = null;
+                try {
+                    InputStream inputStream = (content != null)
+                            ? new ByteArrayInputStream(content.getBytes(StandardCharsets.UTF_8))
+                            : null;
+                    result = HttpUtil.executeUrl(httpMethod, url, httpHeaders, inputStream, contentType, 5000);
+                    logger.debug("[{} {}] {}", httpMethod, url, result);
+                } catch (IOException e) {
+                    // HttpUtil::executeUrl wraps InterruptedException into IOException.
+                    // Unwrap and rethrow it so that we don't retry on InterruptedException
+                    if (e.getCause() instanceof InterruptedException) {
+                        throw (InterruptedException) e.getCause();
+                    }
+
+                    if (e.getCause() instanceof ExecutionException) {
+                        ExecutionException iex = (ExecutionException) e.getCause();
+                        if (iex != null && iex.getCause() instanceof HttpResponseException) {
+                            HttpResponseException hre = (HttpResponseException) iex.getCause();
+                            if (401 == hre.getResponse().getStatus()) {
+                                logger.debug("Perform login");
```

Comment:
better to move this debugging inside the `performlogin` method itself as that method is responsible for such a task.